### ISSUE TITLE
Replay + Scan with initial value emits too many values

### ISSUE THREAD

**Issue by Thorn1089** [ReactiveX/RxJava#4571]
The following test behaves as expected, emitting a single value to the subscriber:

```
    @Test
    public void testExpectedReplayBehavior() {
        final TestScheduler scheduler = new TestScheduler();
        final TestSubject<Integer> subject = TestSubject.create(scheduler);
        final TestSubscriber<Integer> subscriber = new TestSubscriber<>();

        final ConnectableObservable<Integer> sums = subject.scan((a, b) -> a + b).replay(1);
        sums.connect();

        subject.onNext(1);
        subject.onNext(2);
        subject.onNext(3);
        scheduler.triggerActions();

        sums.subscribe(subscriber);

        subscriber.assertValueCount(1);
        subscriber.assertValues(6);
    }
```

However, the second test fails, emitting all of the events from the source subject, rather than replaying the 1 event I requested with `replay(1)`:

```
    @Test
    public void testFlakyReplayBehavior() {
        final TestScheduler scheduler = new TestScheduler();
        final TestSubject<Integer> subject = TestSubject.create(scheduler);
        final TestSubscriber<Integer> subscriber = new TestSubscriber<>();

        final ConnectableObservable<Integer> sums = subject.scan(1, (a, b) -> a + b).replay(1);
        sums.connect();

        subject.onNext(2);
        subject.onNext(3);
        scheduler.triggerActions();

        sums.subscribe(subscriber);

        subscriber.assertValueCount(1);
        subscriber.assertValues(6);
    }
```

Am I missing something nonintuitive about either `scan` or `replay`'s behavior that would reconcile these differences? I don't understand why supplying an initial argument for the accumulator would cause a drastic difference in behavior here.

**Comment by akarnokd** [comment]
Interesting.

`scan(T, Func2)` has an internal buffer that holds onto the intermediate values until the downstream requests. Since `replay` coordinates requests, without a real subscriber, replay doesn't request on its own so `scan` keeps buffering. Once there is a subscriber to `replay`, it requests and now `scan` delivers all buffered values and given the active connection, the `TestSubscriber` receives all 3 values (1, 3, 6). The `scan(Func2)` doesn't buffer and delivers values immediately, filling in the `replay`'s buffer.

Summing up:
1. `TestSubject` doesn't honor backpressure
2. `scan(T, Func2)` buffers too much
3. `replay` doesn't request without `Subscriber`(s)

2 can be considered a bug, 1 should be worked around with proper test emission and 3 is a matter of design and consequences (unlikely to change in 1.x).

**Comment by Thorn1089** [comment]
Thanks for the clarification. I do have a follow-up question around number 1; namely, what should I be doing instead in a test scenario to manually mimic observable emissions if `TestSubject` is not the correct method? Can you clarify what proper test emission would look like here?

**Comment by akarnokd** [comment]
We use `PublishSubject` mostly when there is no real need for a time component but an imperative sequence of emit-then-test code:

``` java
PublishSubject<Integer> ps = PublishSubject.create();

TestSubscriber<Integer> ts = new TestSubscriber<>();

ps.subscribe(ts);

ps.onNext(1);
ts.assertValues(1);

ps.onNext(2);
ts.assertValues(1, 2);

// etc.
```

**Comment by Thorn1089** [comment]
Good to know, thanks!

**Comment by akarnokd** [comment]
I'm closing this issue due to inactivity. If you have further input on the issue, don't hesitate to reopen this issue or post a new one.

**Comment by tgolden-andplus** [comment]
Sorry, can you link the corresponding ticket for the buggy behavior described re: `scan(T, Func2)`?

**Comment by akarnokd** [comment]
There is none. It was unclear if the given scan behavior should be considered a bug or not.
